### PR TITLE
Shared Forms in "Get Blank Forms"

### PR DESCRIPTION
**Pull request by jasonrogena**
Adds a dropdown in the Get Blank Forms page for filtering forms based on owner. Fixes issue #11 

screen_1_resize

Adds the "Show Shared Forms" preference. Fixes issue #6

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java [hunk 1]
```diff
@@ -665,42 +674,151 @@ public void formListDownloadingComplete(HashMap<String, FormDetails> result) {
...
+            //sort how forms will appear in the spinner
+            Collections.sort(userSpinnerValues);
+            ArrayList<String> sortedSpinnerValues = new ArrayList<>();
+
+            SharedPreferences settings =
+                    PreferenceManager.getDefaultSharedPreferences(getBaseContext());
+            String userFormsKey = getResources().getString(R.string.forms_by, settings.getString
+                    (PreferencesActivity.KEY_USERNAME, null));
+
+            if(userSpinnerValues.contains(getResources().getString(R.string.all_forms))) {
```
Comment (pld) [inline collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java]: for consistency add a space after `if`

Comment (pld) [reply collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java]: also DRY up `getResources().getString(R.string.all_forms)` that's here and on the next two lines, but it into a var above

File: collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java [hunk 2]
```diff
@@ -665,42 +674,151 @@ public void formListDownloadingComplete(HashMap<String, FormDetails> result) {
...
+        mFormList.clear();
+
+        ArrayList<String> ids = new ArrayList<String>(forms.keySet());
+        for (int i = 0; i < forms.size(); i++) {
+            String formDetailsKey = ids.get(i);
+            FormDetails details = forms.get(formDetailsKey);
+            HashMap<String, String> item = new HashMap<String, String>();
+            item.put(FORMNAME, details.formName);
+            item.put(FORMID_DISPLAY,
+                    ((details.formVersion == null) ? "" : (getString(R.string.version) + " " + details.formVersion + " ")) +
```
Comment (pld) [inline collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java]: `String.format` would make this cleaner, https://docs.oracle.com/javase/7/docs/api/java/lang/String.html#format(java.lang.String,%20java.lang.Object...)

File: collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java [hunk 3]
```diff
@@ -665,42 +674,151 @@ public void formListDownloadingComplete(HashMap<String, FormDetails> result) {
...
+            item.put(FORMDETAIL_KEY, formDetailsKey);
+            item.put(FORM_ID_KEY, details.formID);
+            item.put(FORM_VERSION_KEY, details.formVersion);
+
+            // Insert the new form in alphabetical order.
+            if (mFormList.size() == 0) {
+                mFormList.add(item);
+            } else {
+                int j;
+                for (j = 0; j < mFormList.size(); j++) {
```
Comment (pld) [inline collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java]: if you insert into the list then sort this happens in O(n + nlogn) = O(n(1 + logn)) = O(nlogn) (upper bound w/mergesort for example), I think your code here will be O(n * n / 2) = O(n^2/2) = O(n^2), so that's significantly worse.

tl;dr, remove all this and sort the list after they've all been inserted

File: collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java [hunk 4]
```diff
@@ -665,42 +674,151 @@ public void formListDownloadingComplete(HashMap<String, FormDetails> result) {
...
+        HashMap<String, HashMap<String, FormDetails>> result = new HashMap<>();
+        result.put(context.getResources().getString(R.string.all_forms), forms);
+
+        SharedPreferences settings =
+                PreferenceManager.getDefaultSharedPreferences(context);
+        boolean showSharedForms = settings.getBoolean(PreferencesActivity.KEY_SHOW_SHARED_FORMS,
+                true);
+        String username = settings.getString(PreferencesActivity.KEY_USERNAME, null);
+
+        for(FormDetails curForm : forms.values()) {
```
Comment (pld) [inline collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java]: missing space here and below too

File: collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java [hunk 5]
```diff
@@ -665,42 +674,151 @@ public void formListDownloadingComplete(HashMap<String, FormDetails> result) {
...
-                        HashMap<String, String> compareMe = mFormList.get(j);
-                        String name = compareMe.get(FORMNAME);
-                        if (name.compareTo(mFormNamesAndURLs.get(ids.get(i)).formName) > 0) {
-                            break;
-                        }
+            HashMap<String, HashMap<String, FormDetails>> userForms =
+                    splitFormsIntoAccounts(getBaseContext(), mFormNamesAndURLs);
+            ArrayList<String> userSpinnerValues = new ArrayList<>(userForms.keySet());
+
+            //sort how forms will appear in the spinner
```
Comment (pld) [inline collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java]: Is there a standard for single line comments, seems like all the others starts with a space and a capitalized first letter

File: collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java [hunk 6]
```diff
@@ -665,42 +674,151 @@ public void formListDownloadingComplete(HashMap<String, FormDetails> result) {
...
+                            new HashMap<String, FormDetails>());
+                }
+
+                result.get(userKey).put(curForm.formID, curForm);
+            } else {
+                Log.e(t, "Could not extract the user that owns the specified form");
+            }
+        }
+
+        if(result.size() == 2) {//forms from only one account
```
Comment (pld) [inline collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java]: why does this guarantee the forms are only from one account?

File: collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java [hunk 7]
```diff
@@ -665,42 +674,151 @@ public void formListDownloadingComplete(HashMap<String, FormDetails> result) {
...
+            //remove the all_forms item to avoid redundant items
+            result.remove(context.getResources().getString(R.string.all_forms));
+        } else if(result.size() == 1 && !showSharedForms) {//only all_forms in map and shared
+            // forms is not turned on. Mean the user doesn't have any forms
+            result.remove(context.getResources().getString(R.string.all_forms));
```
Comment (pld) [inline collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java]: this and the body above on line 796 are the same, why two conditionals instead of a single `if` with an `||`?

File: collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java [hunk 8]
```diff
@@ -665,42 +674,151 @@ public void formListDownloadingComplete(HashMap<String, FormDetails> result) {
...
+
+    /**
+     * This method extracts the Ona user that owns a form from the form's Download URL
+     *
+     * @param formDownloadUrl   The download URL for the form (obtained from /formList endpoint)
+     *
+     * @return  The Ona User that owns the form or null if something goes wrong
+     */
+    public static String getOnaUser(String formDownloadUrl) {
+        //https://odk.ona.io/yri/forms/129132/form.xml
```
Comment (pld) [inline collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java]: this is an example? mention that

File: collect_app/src/main/java/org/odk/collect/android/preferences/AdminPreferencesActivity.java [hunk 9]
```diff
@@ -81,6 +81,8 @@
     public static String KEY_AUTOSEND_WIFI = "autosend_wifi";
     public static String KEY_AUTOSEND_NETWORK = "autosend_network";
 
+	public static String KEY_SHOW_SHARED_FORMS = "show_shared_forms";
```
Comment (pld) [inline collect_app/src/main/java/org/odk/collect/android/preferences/AdminPreferencesActivity.java]: remove extra spaces

File: collect_app/src/test/java/org/odk/collect/android/activities/FormDownloadListTest.java [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    public void testGetOnaUser() {
+        String username1 = "dfsDFE12_fsd";
+        //test a normal url
+        String url1 = "https://odk.ona.io/"+username1+"/forms/129132/form.xml";
+        Assert.assertTrue(FormDownloadList.getOnaUser(url1).equals(username1));
+
+        //test http url
+        String url2 = "http://odk.ona.io/"+username1+"/forms/129132/form.xml";
+        Assert.assertTrue(FormDownloadList.getOnaUser(url2).equals(username1));
+
```
Comment (pld) [inline collect_app/src/test/java/org/odk/collect/android/activities/FormDownloadListTest.java]: let's add a test for "http://ona.io/"+username1+"/forms/129132/form.xml", this should be valid

**Review by pld — CHANGES_REQUESTED** [review]

File: collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java (hunk 5 above)
Comment (jasonrogena) [reply collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java]: There seems to be none for this project, but will do

File: collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java (hunk 3 above)
Comment (jasonrogena) [reply collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java]: Will do; will probably also push upstream (this is part of a code-block inherited from ODK Collect, I moved it into a method just for cleanliness)

File: collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java (hunk 6 above)
Comment (jasonrogena) [reply collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java]: All forms are initially all put in the first item in the map (called "all_forms"). Cases where you end up with two items in the map after splitting forms by accounts means that there's only one account

File: collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java (hunk 3 above)
Comment (pld) [reply collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java]: ah cool

File: collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java (hunk 6 above)
Comment (pld) [reply collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java]: cool, can you add a note to that affect so people know why this works

**Review by pld — APPROVED** [review]

**Comment by skambo** [comment]
@jasonrogena Looks good, I'm able to sort forms by project plus the settings option to show forms shared with you on Ona works.
